src/ddmd/intrange.d

```diff
@@ -608,73 +603,153 @@ struct IntRange
     {
-        // Not yet implemented
-        assert(0);
+        return IntRange(imin - rhs.imax, imax - rhs.imin);
     }
 
-    R opBinary(string op : "*")(R rhs)
+    IntRange opBinary(string op : "*")(IntRange rhs)
     {
-        // Not yet implemented
-        assert(0);
+        // [a,b] * [c,d] = [min (ac, ad, bc, bd), max (ac, ad, bc, bd)]
+        SignExtendedNumber[4] bdy;
+        bdy[0] = imin * rhs.imin;
+        bdy[1] = imin * rhs.imax;
+        bdy[2] = imax * rhs.imin;
+        bdy[3] = imax * rhs.imax;
+        return IntRange.fromNumbers4(bdy.ptr);
     }
 
-    R opBinary(string op : "/")(R rhs)
+    IntRange opBinary(string op : "/")(IntRange rhs)
     {
-        // Not yet implemented
-        assert(0);
+        // Handle divide by 0
+        if (rhs.imax.value == 0 && rhs.imin.value == 0) return widest();
+
+        // Don't treat the whole range as divide by 0 if only one end of a range is 0.
+        // Issue 15289
+        if (rhs.imax.value == 0) rhs.imax.value--;
+        else if(rhs.imin.value == 0) rhs.imin.value++;
+
+        if (!imin.negative && !imax.negative && !rhs.imin.negative && !rhs.imax.negative)
+        {
+            auto res = IntRange(imin / rhs.imax, imax / rhs.imin);
+            return res;
+        }
+        else
+        {
+            // [a,b] / [c,d] = [min (a/c, a/d, b/c, b/d), max (a/c, a/d, b/c, b/d)]
+            SignExtendedNumber[4] bdy;
+            bdy[0] = imin / rhs.imin;
+            bdy[1] = imin / rhs.imax;
+            bdy[2] = imax / rhs.imin;
+            bdy[3] = imax / rhs.imax;
+
+            auto res = IntRange.fromNumbers4(bdy.ptr);
+            //printf("%u %u %d\n", res.imin.value, res.imax.value);
+            return res;
+        }
     }
 
-    R opBinary(string op : "%")(R rhs)
+    IntRange opBinary(string op : "%")(IntRange rhs)
     {
-        // Not yet implemented
-        assert(0);
+        IntRange irNum = this;
+        IntRange irDen = rhs.absNeg();
+
+        /*
+         due to the rules of D (C)'s % operator, we need to consider the cases
+         separately in different range of signs.
+
+             case 1. [500, 1700] % [7, 23] (numerator is always positive)
+                 = [0, 22]
+             case 2. [-500, 1700] % [7, 23] (numerator can be negative)
+                 = [-22, 22]
+             case 3. [-1700, -500] % [7, 23] (numerator is always negative)
+                 = [-22, 0]
+
+         the number 22 is the maximum absolute value in the denomator's range. We
+         don't care about divide by zero.
+         */
+
+        irDen.imin = irDen.imin + SignExtendedNumber(1);
+        irDen.imax = -irDen.imin;
+
+        if (!irNum.imin.negative)
+            irNum.imin.value = 0;
+        else if (irNum.imin < irDen.imin)
+            irNum.imin = irDen.imin;
+
+        if (irNum.imax.negative)
+        {
+            irNum.imax.negative = false;
+            irNum.imax.value = 0;
```

Comment:
Since this is code added by this PR, it really could use a test case. VRP is so easy to get wrong, having a thorough test suite for code changes is important.